### PR TITLE
WOL widget rewrite for new dashboard

### LINKED ISSUE #4114: os-wol widget does not appear in 24.7
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

- [x ] I have read the contributing guide lines at https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md
- [x ] I have searched the existing issues, open and closed, and I'm convinced that mine is new.
- [x ] The title contains the plugin to which this issue belongs

**Describe the bug**
After upgrade to the 24.7 WakeOnLan widget is not available anymore.

**To Reproduce**
1. Go to Dashboard
2. Click on "Add Widget"
3. Open available list
4. See that Wake On Lan is no longer available

**Expected behavior**
Widget is available for selection and operates in similar manner as in pre 24.7 versions.

I'm aware that MVC changed with the 24.7 release and widget requires a rewrite, but as I'm unfamiliar with the code I'm unable to contribute a fix. Hopefully someone else can help.

### PR DESCRIPTION
**Pull request by m3e-g**
This is my attempt on rewrite for the WOL widget. 

image

It's functionality is mostly the same as the old one, with the exception of ARP check which is now done via the API call.

As it's my first time with the JS, any inputs are welcome.

This should fix #4114.

### CODE REVIEW

File: net/wol/Makefile [hunk 1]
```diff
@@ -1,6 +1,6 @@
 PLUGIN_NAME=		wol
-PLUGIN_VERSION=		2.4
-PLUGIN_REVISION=	2
+PLUGIN_VERSION=		3.0
+PLUGIN_REVISION=	1
```
Comment (fichtner) [inline net/wol/Makefile:3]:
```suggestion
PLUGIN_VERSION=		2.5
```

I think 2.5 should be exciting enough :)

File: net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+<metadata>
+    <wol>
+        <filename>WakeOnLan.js</filename>
+        <endpoints>
+           <endpoint>/api/wol/wol/*</endpoint>
+           <endpoint>/diagnostics/interface/*Arp</endpoint>
```
Comment (fichtner) [inline net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml:6]: It's better to list everything here vs. using `*Arp` for future robustness.

File: net/wol/src/opnsense/www/js/widgets/WakeOnLan.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ * AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
+ * OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import BaseTableWidget from "./BaseTableWidget.js";
```
Comment (fichtner) [inline net/wol/src/opnsense/www/js/widgets/WakeOnLan.js:27]: that's the old old style. I think for now we can remove this and I'll fix a backport if we need it.

```suggestion
```

File: net/wol/src/opnsense/www/js/widgets/WakeOnLan.js [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+  checkActive(arp_list, mac, intf) {
+    const arp = arp_list.find((obj) => obj.mac === mac);
+    if (arp === undefined) {
+      return 0;
+    } else {
+      return (arp.expired === false && arp.intf_description === intf);
+    }
+  }
+
```
Comment (fichtner) [inline net/wol/src/opnsense/www/js/widgets/WakeOnLan.js:98]:
```suggestion
```

**Comment by fichtner** [comment]
@m3e-g thanks, taking a look at the JS code tomorrow

File: net/wol/src/opnsense/www/js/widgets/WakeOnLan.js (hunk 3 above)
Comment (m3e-g) [reply net/wol/src/opnsense/www/js/widgets/WakeOnLan.js:27]: Without this import, on the latest stable (24.7.1) I have this error:

```
[ReferenceError: BaseTableWidget is not defined
    at https://opn-m3e/ui/js/widgets/WakeOnLan.js?t=…]
```
image

Comment (fichtner) [reply net/wol/src/opnsense/www/js/widgets/WakeOnLan.js:27]: 21.7.1 & .2 actually require an importmap-based statement. 24.7 required your current statement. 24.7.3 will likely require no statement which is also the version we need to merge to the master branch

File: net/wol/Makefile (hunk 1 above)
Comment (m3e-g) [reply net/wol/Makefile]: I've been wondering about this one. On one hand, the plugin change was not been big enough to justify big version bump, but on the other I wanted to indicate that this is different code base. 
Anyway I've changed the version according to your suggestion.

File: net/wol/src/opnsense/www/js/widgets/WakeOnLan.js (hunk 3 above)
Comment (m3e-g) [reply net/wol/src/opnsense/www/js/widgets/WakeOnLan.js]: Thanks for the explanation. I've committed the version without the import statement

File: net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml (hunk 2 above)
Comment (m3e-g) [reply net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml]: Fixed in 1d700fb39e8a3bfc4fe7632654402c64f66fa2d5

File: net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+      <endpoint>/diagnostics/interface/getArp</endpoint>
+    </endpoints>
+    <translations>
+      <title>WakeOnLan</title>
+      <msg_empty_wol>No saved WOL addresses</msg_empty_wol>
+      <h_device>Device</h_device>
+      <h_interface>Interface</h_interface>
+      <h_status>Status</h_status>
+    </translations>
+  </wol>
```
Comment (fichtner) [inline net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml:16]:
```suggestion
  <wol>
      <filename>WakeOnLan.js</filename>
      <endpoints>
          <endpoint>/api/wol/wol/searchHost</endpoint>
          <endpoint>/api/wol/wol/set</endpoint>
          <endpoint>/diagnostics/interface/getArp</endpoint>
      </endpoints>
      <translations>
          <title>WakeOnLan</title>
          <msg_empty_wol>No saved WOL addresses</msg_empty_wol>
          <h_device>Device</h_device>
          <h_interface>Interface</h_interface>
          <h_status>Status</h_status>
    </translations>
</wol>
```

indent 4 spaces

File: net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+          <endpoint>/diagnostics/interface/getArp</endpoint>
+      </endpoints>
+      <translations>
+          <title>WakeOnLan</title>
+          <msg_empty_wol>No saved WOL addresses</msg_empty_wol>
+          <h_device>Device</h_device>
+          <h_interface>Interface</h_interface>
+          <h_status>Status</h_status>
+    </translations>
+</wol>
```
Comment (fichtner) [inline net/wol/src/opnsense/www/js/widgets/Metadata/WakeOnLan.xml:16]:
```suggestion
    <wol>
        <filename>WakeOnLan.js</filename>
        <endpoints>
            <endpoint>/api/wol/wol/searchHost</endpoint>
            <endpoint>/api/wol/wol/set</endpoint>
            <endpoint>/diagnostics/interface/getArp</endpoint>
        </endpoints>
        <translations>
            <title>WakeOnLan</title>
            <msg_empty_wol>No saved WOL addresses</msg_empty_wol>
            <h_device>Device</h_device>
            <h_interface>Interface</h_interface>
            <h_status>Status</h_status>
        </translations>
    </wol>
```

whoops :)

**Comment by fichtner** [comment]
Merging for further testing, thanks a lot!

**Comment by fichtner** [comment]
@m3e-g can you check https://github.com/opnsense/core/commit/8b5a9a17 plus https://github.com/opnsense/plugins/commit/40d28ac149

Noticed that the device wasn't showing as online due to a mismatch in key vs. value lookup on the arp end.

**Comment by m3e-g** [comment]
@fichtner I've checked this change and now it no longer works on my side ;)

I'm using bridge interface, which called "bridge0", and interface description is "LAN". 
`wol/searchHost` is returning the 
```javascript
[{ ... identifier: "lan", interface: "LAN" ... }]
```
so the comparison in the checkActive fails. 

Is the lowercase conversion expected? If so then it can be handled within the widget in above function.
If not then maybe `searchHost` could return the "intf" like the `getArp`? 
Sadly I have even less experience with the backend structure of the plugin...

Icon changing on click is a nice touch for giving user the required feedback.

Thank you very much for working on it!

**Comment by fichtner** [comment]
Can you give an overview of your interfaces and assigned device names?

**Comment by fichtner** [comment]
Maybe we just want to omit checking the interface name so it shows as online regardless of where it is although wake may or make not work?

**Comment by m3e-g** [comment]
> Can you give an overview of your interfaces and assigned device names?

```
igc0: 
igc1: description: OPT1 (opt1)
igc2: description: OPT2 (opt2)
igc3: description: OPT3 (opt3)
igc4: description: OPT4 (opt4)
igc5: description: OPT5 (opt5)
lo0: 
enc0: 
pfsync0: 
maxupd: 
pflog0:
pppoe0: description: WAN (wan)
bridge0: 
        description: LAN (lan)
        member: igc5
        member: igc4
        member: igc3
        member: igc2
        member: igc1
        groups: bridge
tailscale0: 
```

**Comment by m3e-g** [comment]
> > [{ ... identifier: "lan", interface: "LAN" ... }]
> 
> This appears correct unless you changed the interface assignment or have "LAN" there in it twice. I'm not sure how arp works in terms of brige0, but I think it only worked by accident showing active.

`arp` (command) does not show the interface descriptions, but the `getArp` api does return both interface name and description:

```json
{
  "mac": "xx:xx:xx:xx:xx:xx:xx:xx",
  "ip": "192.168.XX.YY.",
  "intf": "bridge0",
  "expired": false,
  "expires": 405,
  "permanent": false,
  "type": "bridge",
  "manufacturer": "ASUSTek COMPUTER INC.",
  "hostname": "",
  "intf_description": "LAN"
}
```

Because the `wol/searchHost` had been returning only the interface name I've been comparing it with `intf_description`. This is not ideal, and probably should be handled in the wol API.

Widget did not have interface check in the previous version of the plugin, but as it's been a free information stored in the `getArp` results anyway I've been think it would be nice to know if the device had been brought up on correct interface. 
Some of the embedded devices have dummy/duplicated MACs, so why not.

As I don't know how to properly modify the backend side of the plugin, we can remove the interface check and have the original functionality for the time being

**Comment by fichtner** [comment]
this one actually, now I can roll back the change. sorry

https://github.com/opnsense/core/commit/c296ddea

**Comment by fichtner** [comment]
I rolled it back. Can you try latest master?

**Comment by fichtner** [comment]
https://github.com/opnsense/core/commit/e2bd3c33d

    # opnsense-patch e2bd3c33d

This change is going into 24.7.3 anyway and we don't have to mess with import statements anymore.


Cheers,
Franco

**Comment by fichtner** [comment]
Sorry lots of patches, but it will all make sense on 24.7.3 with the widget released :)

**Comment by m3e-g** [comment]
Here's what I did:

```
opnsense-patch c296ddea
opnsense-patch e2bd3c33d
configctl webgui restart
```

It does work without the import statement and the widget matches the ARP entry correctly.
Thanks!

**Comment by fichtner** [comment]
Great, I'll release the widget and everything else discussed with 24.7.3.

Thanks again for the widget!


Cheers,
Franco

**Pull request merged**